src/libAtomVM/nifs.c

```diff
@@ -2466,3 +2478,161 @@ static term nif_console_print(Context *ctx, int argc, term argv[])
     }
     return OK_ATOM;
 }
+
+static char b64_table[64] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+
+// per https://tools.ietf.org/rfc/rfc4648.txt
+
+static term nif_atomvm_b64_encode(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    term src = argv[0];
+
+    VALIDATE_VALUE(src, term_is_binary);
+
+    size_t src_size = term_binary_size(src);
+    if (src_size == 0) {
+        return src;
+    }
+    uint8_t *src_buf = malloc(src_size);
+    if (IS_NULL_PTR(src_buf)) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    memcpy(src_buf, term_binary_data(src), src_size);
+    uint8_t *src_pos = src_buf;
+    size_t dst_size = (4 * src_size) / 3;
+    size_t pad = 0;
+    switch (src_size % 3) {
+        case 0:
+            break;
+        case 1:
+            pad = 2;
+            dst_size++;
+            break;
+        case 2:
+            pad = 1;
+            dst_size++;
+            break;
+    }
+    if (UNLIKELY(memory_ensure_free(ctx, term_binary_data_size_in_terms(dst_size + pad) + BINARY_HEADER_SIZE) != MEMORY_GC_OK)) {
```

Comment:
in case we avoid copying memory, we should call memory_ensure_free before term_binary_data(src).

Reply:
Ah right, that makes sense.  Wasn't thinking I could check first before GC

Reply:
If you do `memory_ensure_free` as first thing, and you access src later it will never change again so it is completely safe (since you are not going to call `memory_ensure_free` again).